app/code/Magento/Inventory/Model/SourceItem/Validator/ConsistencyValidator.php

```diff
@@ -0,0 +1,72 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\Inventory\Model\SourceItem\Validator;
+
+use Magento\Framework\Validation\ValidationResult;
+use Magento\Framework\Validation\ValidationResultFactory;
+use Magento\Inventory\Model\OptionSource\SourceItemStatus;
+use Magento\InventoryApi\Api\Data\SourceItemInterface;
+
+/**
+ * Check that if product quantity <= 0 status is not "In Stock".
+ */
+class ConsistencyValidator implements SourceItemValidatorInterface
+{
+    /**
+     * @var ValidationResultFactory
+     */
+    private $validationResultFactory;
+
+    /**
+     * @var SourceItemStatus
+     */
+    private $sourceItemStatus;
+
+    /**
+     * @param ValidationResultFactory $validationResultFactory
+     * @param SourceItemStatus $sourceItemStatus
+     */
+    public function __construct(
+        ValidationResultFactory $validationResultFactory,
+        SourceItemStatus $sourceItemStatus
+    ) {
+        $this->validationResultFactory = $validationResultFactory;
+        $this->sourceItemStatus = $sourceItemStatus;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function validate(SourceItemInterface $source): ValidationResult
+    {
+        $status = $source->getStatus();
+        $quantity = $source->getQuantity();
+        $statusOptions = $this->sourceItemStatus->toOptionArray();
+        $label = '';
+        foreach ($statusOptions as $statusOption) {
```

Comment:
Use array_column instead of `foreach`
And move `under` if condition (label is needed only if we have error)

Reply:
Fixed